rse-bash.Rmd

```diff
@@ -1256,801 +1176,895 @@ $ ls lengths.txt
-> the output of a command that operates on a file
-> to the same file. For example:
+> What happens if a command is supposed to process a file
+> but we don't give it a filename?
+> For example, what if we type:
 >
 > ```shell
-> $ sort -n lengths.txt > lengths.txt
+> $ wc -l
+> ```
+>
+> but don't type `*.csv` (or anything else) after the command?
+> Since it doesn't have any filenames,
+> `wc` assumes it is supposed to read from the keyboard,
+> so it waits for us to type in some data.
+> It doesn't tell us this:
+> it just sits and waits.
+>
+> This mistake can be hard to spot,
+> particularly if we put the filename at the end of the pipeline:
+>
+> ```shell
+> $ wc -l | sort bellambi_temp.csv
 > ```
 >
-> Doing something like this may give you
-> incorrect results and/or delete
-> the contents of `lengths.txt`.
+> In this case,
+> `sort` ignores standard input and reads the data in the file,
+> but `wc` still just sits there waiting for input.
+>
+> If we make this mistake,
+> we can end the program by typing <kbd>Ctrl</kbd>+<kbd>C</kbd>.
+> We can also use this to interrupt programs that are taking a long time to run
+> or are trying to connect to a website that isn't responding.
 
-If you think this is confusing,
-you're in good company:
-even once you understand what `wc`, `sort`, and `head` do,
-all those intermediate files make it hard to follow what's going on.
-We can make it easier to understand by running `sort` and `head` together:
+Just as we can redirect standard output with `>`,
+we can connect standard input to a file using `<`.
+In the case of a single file,
+this has the same effect as providing the file's name to the command:
 
 ```shell
-$ sort -n lengths.txt | head -n 1
+$ wc < bellambi_temp.csv
 ```
 
 ```text
-  9  methane.pdb
+    8314    8316  184203
 ```
 
-The vertical bar, `|`, between the two commands is called a **pipe**.
-It tells the shell that we want to use
-the output of the command on the left
-as the input to the command on the right.
-The computer might create a temporary file if it needs to,
-or copy data from one program to the other in memory,
-or something else entirely;
-we don't have to know or care.
+If we try to use redirection with a wildcard,
+though,
+the shell *doesn't* concatenate all of the matching files:
+
+```shell
+$ wc < *.csv
+```
+
+```text
+-bash: *.csv: ambiguous redirect
+```
+
+It also doesn't print the error message to standard output,
+which we can prove by redirecting:
+
+```shell
+$ wc < *.csv > all.txt
+```
+
+```text
+-bash: *.csv: ambiguous redirect
+```
+
+```shell
+$ cat all.txt
+```
+
+```text
+cat: all.txt: No such file or directory
+```
+
+Every process has a second output channel called [standard error][standard-error] (or [stderr][stderr]).
+Programs use it for error messages
+so that their attempts to tell us something has gone wrong don't vanish silently into an output file.
+There are ways to redirect standard error,
+but doing so is almost always a bad idea.
```

Comment:
Isn't it quite useful for logging output from a command (e.g. with tee or straight to file)? Bash somewhat recently introduced `|&` as an abbreviation for `2>&1 |`, so I would believe it sees a fair amount of use to justify a shortcut.